digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/LaneUtils.scala

```diff
@@ -127,78 +135,92 @@ object LaneUtils {
+                                linkLength: Double): Option[(Double, Double)] = {
+    val (linkAddressM, linkMValue) = // Determine total address length and m value within link
+      addressesOnLink.foldLeft(0L, 0.0){ (result, address) =>
+        (result._1 + address.endAddressM - address.startAddressM, result._2 + address.endMValue - address.startMValue)
       }
-      else {
-        ( road.startMValue, road.endMValue )
 
+    // Viite uses frozen links, so determine new measure that matches current link length
+    val viiteMeasure = linkMValue / linkAddressM
+    val adjustedMeasure = linkLength / linkAddressM
+
+    val startAndEndValues = addressesOnLink.flatMap { road =>
+      val startMValue = road.startMValue / viiteMeasure * adjustedMeasure
+      val endMValue = road.endMValue / viiteMeasure * adjustedMeasure
+      val startDifferenceM = (laneRoadAddressInfo.initialDistance - road.startAddressM) * adjustedMeasure
+      val endDifferenceM = (road.endAddressM - laneRoadAddressInfo.endDistance) * adjustedMeasure
+      val towardsDigitizing = road.sideCode.getOrElse(SideCode.TowardsDigitizing) == SideCode.TowardsDigitizing
+
+      val startPoint = // Calculated start point for cases when lane does not start from the start of road address
+        if (towardsDigitizing && startDifferenceM > 0) startMValue + startDifferenceM
+        else if (!towardsDigitizing && endDifferenceM > 0) startMValue + endDifferenceM
+        else startMValue
+      val endPoint = // Calculated end point for cases when lane ends before the end of road address
+        if (towardsDigitizing && endDifferenceM > 0) endMValue - endDifferenceM
+        else if (!towardsDigitizing && startDifferenceM > 0) endMValue - startDifferenceM
+        else endMValue
+
+      // Determine if what endpoint values to use or if road address should be skipped
+      // If adjusted start or end point is used and road side code is againstDigitising, the opposite value is adjusted
+      road.roadPart match {
+        case part if part > laneRoadAddressInfo.initialRoadPartNumber && part < laneRoadAddressInfo.endRoadPartNumber =>
+          Some(road.startMValue, road.endMValue)
+
+        case part if part == laneRoadAddressInfo.initialRoadPartNumber && part == laneRoadAddressInfo.endRoadPartNumber =>
+          if (!(road.endAddressM > laneRoadAddressInfo.initialDistance && road.startAddressM < laneRoadAddressInfo.endDistance))
+            None
+          else if (road.startAddressM <= laneRoadAddressInfo.initialDistance && road.endAddressM >= laneRoadAddressInfo.endDistance)
+            Some( startPoint, endPoint )
+          else if (road.startAddressM <= laneRoadAddressInfo.initialDistance)
+            if (towardsDigitizing) Some( startPoint, endMValue ) else Some( startMValue, endPoint )
+          else if (road.endAddressM >= laneRoadAddressInfo.endDistance)
+            if (towardsDigitizing) Some( startMValue, endPoint ) else Some( startPoint, endMValue )
```

Comment:
Tää toistuu jonkin verran pienillä eroilla.
if (towardsDigitizing) Some( startMValue, endPoint ) else Some( startPoint, endMValue 

Voisi yrittää vähentää toistoa.

Reply:
Onko tarkempaa ajatusta miten? Esim tässä ensimmäisessa kohdassa valinta on lyhyempi kuin linkin arvo joten vain toista loppupistettä pitää säätää. Jos osoite on digitointisuuntaan niin sitten säädettäisiin alkupistettä ja jos se on päin vastoin digitointisuuntaan niin sitten säädetään loppupistettä.
Toinen vaihtoehto olisi lisätä toi digitointi suunnan tarkastus tonne ylempään ehtoon, mutta ajattelin, se tekisi tuosta vaikeammin luettavan esim:
else if ((road.startAddressM <= laneRoadAddressInfo.initialDistance && towardsDigitizing) || (road.endAddressM >= laneRoadAddressInfo.endDistance && !towardsDigitizing))  Some( startPoint, endMValue )

Reply:
kai se voi olla noin, en ihan heti keksi miten pääsis tosta duplikaatista eroon.